---
title: '`r paste("Fusion Sample Report", params$samplename)`'
author: "EGLH Cancer Team"
date: '`r format(Sys.time(), "Last modified: %d %b %Y")`'
params:
  qcmetrics:
  qcgenecov:
  capturegenes:
  qcexonscov:
  fusionsabridgedcoding:
  ensbsqlite:
  bam:
  samplename:
  sexphenotype:
  targetlist:
  ref_annot:
  mane:
  chimerkb:
output:
  html_document:
    code_folding: hide
    toc: true
    toc_float: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE ,warning = FALSE, message = FALSE)
```

```{r, warning=FALSE}
# load libraries
library(DT)
library(chimeraviz)
library(tidyr)
install.packages("readxl")
library("readxl")
library(dplyr)
install.packages("gt")
library(gt)
```


## 1. Sample Info

```{r}
fusion_abdriged <- read.csv("MB2002.FusionInspector.fusions.abridged.tsv.coding_effect", sep = "\t")
fusion_abdriged$X.FusionName <- gsub("--", "::",fusion_abdriged$X.FusionName)
```

### 1.1 Sample Meta Data

Sample ID: `r params$samplename`  
Gender: `r params$sexphenotype`  
Age: NA  
Tumour Type: NA  
Target list: `r params$targetlist`  
Found fusions: `r fusion_abdriged$X.FusionName`  

### 1.2 Sample QC

GTEx v8 QC thresholds, not normals are any metrics where:

* < 10 million mapped reads;
* read mapping rate < 0.2;
* intergenic mapping rate > 0.3;
* rRNA mapping rate > 0.3.


```{r}
metrics <- read.csv("MB2002.star.bam.metrics.tsv", sep = "\t", header = F)
colnames(metrics) <- c("QC_metric", "Value")
metrics_filtered <- metrics[which(metrics$QC_metric %in% c("Mapping Rate", "Expression Profiling Efficiency", "Exonic Rate",
                                                    "Read Length", "Genes Detected", "Median Exon CV", "Exon CV MAD",
                                                    "Intronic Rate","Intergenic Rate","Intragenic Rate", "rRNA Rate",
                                                    "Duplicate Reads", "Mapped Reads", "Mapped Unique Reads", "rRNA Reads", "Total Reads")), ]
metrics_filtered$Status <- rep(NA, nrow(metrics_filtered))
# check the 4 QC_metrics thresholds are met
options(scipen=99)
metric_threshold_df <- data.frame(c("Mapped Reads", "Mapping Rate", "Intergenic Rate", "rRNA Rate"),
                                  c(10000000, 0.2 ,0.2, 0.3),
                                  c(">", ">", "<", "<"))
colnames(metric_threshold_df) <- c("metric", "value", "pass_condition")

for (i in 1:nrow(metric_threshold_df)){
  metric = metric_threshold_df[i, "metric"]
  threshold = metric_threshold_df[i, "value"]
  condition = metric_threshold_df[i, "pass_condition"]
  if (condition == ">" ){
    if ( metrics_filtered[which(metrics_filtered$QC_metric == metric),"Value"] > threshold ){
      metrics_filtered[which(metrics_filtered$QC_metric == metric),"Status"] <- "PASS"
    } else {
      metrics_filtered[which(metrics_filtered$QC_metric == metric),"Status"] <- "FAIL"
    }
  }
  else {
    if ( metrics_filtered[which(metrics_filtered$QC_metric == metric),"Value"] < threshold ){
      metrics_filtered[which(metrics_filtered$QC_metric == metric),"Status"] <- "PASS"
    } else {
      metrics_filtered[which(metrics_filtered$QC_metric == metric),"Status"] <- "FAIL"
    }
  }
}

metrics_filtered <- metrics_filtered[order(metrics_filtered$Status),]
datatable(metrics_filtered)
```

## 2. RNA Coverage 

### 2.1 Coverage by Gene


```{r coverage}
rnaseqc_coverage <- read.csv(params$qcgenecov, sep = "\t")
# get the genes on the capture
capture_genes <- read.csv(params$capturegenes, header = F)
rnaseqc_coverage_hgnc <- rnaseqc_coverage[which(rnaseqc_coverage$hgnc_symbol %in% capture_genes$V1),]
max_cov <- as.numeric(rnaseqc_coverage_hgnc[order(rnaseqc_coverage_hgnc$coverage_mean, decreasing = T),][1,"coverage_mean"])
max_gene <-  rnaseqc_coverage_hgnc[order(rnaseqc_coverage_hgnc$coverage_mean, decreasing = T),][1,"hgnc_symbol"]                 
```

```{r}
ggplot(rnaseqc_coverage_hgnc, aes(rnaseqc_coverage_hgnc$coverage_mean)) +
  geom_boxplot() + 
  theme(
         panel.background = element_rect(fill='transparent'),
         plot.background = element_rect(fill='transparent', color=NA),
         panel.grid.major = element_blank(),
         panel.grid.minor = element_blank(),
         legend.background = element_rect(fill='transparent'),
         legend.box.background = element_rect(fill='transparent')
       ) +
  labs(x = "Mean Coverage", title = "All genes")

ggplot(rnaseqc_coverage_hgnc, aes(rnaseqc_coverage_hgnc$coverage_mean)) +
  geom_boxplot() + 
  theme(
         panel.background = element_rect(fill='transparent'),
         plot.background = element_rect(fill='transparent', color=NA),
         panel.grid.major = element_blank(),
         panel.grid.minor = element_blank(),
         legend.background = element_rect(fill='transparent'),
         legend.box.background = element_rect(fill='transparent')
       ) +
  xlim(0, 50000) +
  labs(x = "Mean Coverage", title = "All genes with mean coverages < 50000")

```


```{r}
datatable(rnaseqc_coverage_hgnc, editable = TRUE)
```


### 2.2 Coverage by Exon

```{r econ_coverage}
rnaseqc_exons <- read.csv(params$qcexonscov, sep = "\t")
rnaseqc_exons <- separate(data = rnaseqc_exons, col = Exon.ID, into = c("Gene_ID", "Exon"), sep = "_")
# get the genes on the capture
rnaseqc_exonse_hgnc <- rnaseqc_exons[which(rnaseqc_exons$Gene_ID %in% rnaseqc_coverage_hgnc$ENSG_id),]
rnaseqc_exonse_hgnc$hgnc_symbol <- rnaseqc_coverage_hgnc$hgnc_symbol[match(rnaseqc_exonse_hgnc$Gene_ID, rnaseqc_coverage_hgnc$ENSG_id)]
rnaseqc_exonse_hgnc$hgnc_id<- rnaseqc_coverage_hgnc$hgnc_id[match(rnaseqc_exonse_hgnc$Gene_ID, rnaseqc_coverage_hgnc$ENSG_id)]
rnaseqc_exonse_hgnc <- rnaseqc_exonse_hgnc[,c("hgnc_id", "hgnc_symbol", "Gene_ID", "Exon", "Exon.CV")]
```

```{r}
datatable(rnaseqc_exonse_hgnc, editable = TRUE)
```

## 3. Fusion Prediction

### 3.1 Fusion Prediction {.tabset .tabset-dropdown}

#### Columm names defintions

    FusionName: name of the fusion as geneA--geneB  
    JunctionReadCount: number of split RNA-Seq reads that map and define the fusion breakpoint.  
    SpanningFragCount: number of paired-end reads that span the fusion breakpoint but the reads do not directly overlap the breakpoint.  
    LeftGene: identifier of the gene represented by the left section of the fusion transcript.  
    LeftLocalBreakpoint: position of the left fusion breakpoint in the context of the FusionInspector mini-fusion-gene contig.  
    LeftBreakpoint: position of the left fusion breakpoint in the context of the genome.  
    RightGene:  identifier of the gene represented by the right section of the fusion transcript.  
    RightLocalBreakpoint: position of the right fusion breakpoint in the context of the FusionInspector min-fusion-gene contig.  
    RightBreakpoint: position of the right fusion breakpoint in the context of the genome.  
    SpliceType: category of support at the fusion breakpoint: {
                ONLY_REF_SPLICE: fusion breakpoint occurs at reference (known) splice junctions.
                INCL_NON_REF_SPLICE: fusion breakpoint occurs at a breakpoint that does not involve all reference (known) exon junctions.
                NO_JUNCTION_READS_IDENTIFIED: only spanning fragments support the fusion. (can only happen if --min_junction_reads is set to zero).
           }  
    LargeAnchorSupport:  YES|NO, indicates whether there are at least 25 aligned bases on each side of the fusion breakpoint.  
    NumCounterFusionLeft:  Number of RNA-Seq fragments supporting the alternative non-fused allele at the left junction breakpoint.   
    NumCounterFusionRight:  Number of RNA-Seq fragments supporting the alternative non-fused allele at the right junction breakpoint.  
    FAR_left:  Left fusion allelic ratio  
    FAR_right:  Right fusion allelic ratio  
    LeftBreakDinuc: the genomic dinucleotides found at the left breakpoint (putative splice site if splicing is involved).  
    LeftBreakEntropy: entropy calculation for the 15 bases immediately upstream from the fusion junction.  
    RightBreakDinuc:  the genomic dinucleotides found at the right breakpoint (putative splice site if splicing is involved).  
    RightBreakEntropy: entropy calculation for the 15 bases immediately downstream from the fusion junction.  
    FFPM: normalized measure of the quantity of RNA-Seq fragments supporting the fusion event as:  fusion fragments per total million RNA-Seq fragments.  

```{r}
fusion_abdriged <- read.csv(params$fusionsabridgedcoding, sep = "\t")
fusion_abdriged <- fusion_abdriged[,c(1:30,34,35,31,32,33)]
datatable(fusion_abdriged)


```


### 3.2 Exon breakpoint annotation


```{r Exons table}
mane <- read.csv(params$mane, sep = "\t")
fusion_abdriged <- read.csv(params$fusionsabridgedcoding, sep = "\t")
fusion_abdriged <- fusion_abdriged[,c(1:30,34,35,31,32,33)]

ref_annot <- read.delim(params$ref_annot, sep = "\t", skip = 5, header = F)
colnames(ref_annot) <- c("chr", "source", "feature", "start", "end", "score", "strand", "frame", "attribute")
# filter the reference annotation file to ensembl source and exons only
ref_annot_exons <- ref_annot[which(ref_annot$feature == "exon"), ]

# get the genes names to filter the ref annotation to genes of interest
genes <- unlist(strsplit(fusion_abdriged$X.FusionName, "--"))
ref_annot_df <- as.data.frame(matrix(NA, ncol = ncol(ref_annot_exons), nrow = 0))
colnames(ref_annot_df) <- colnames(ref_annot_exons)

for(g in genes){
  ref_annot_exons_filtered <- ref_annot_exons[grep(paste(g,";", sep=""), ref_annot_exons$attribute, fixed = T) , ]
  ref_annot_df <- rbind(ref_annot_df, ref_annot_exons_filtered)
}

# ensembl attribute
transcript_attributes <- c("gene_id", "transcript_id", "gene_type", "gene_name",
                           "transcript_type", "transcript_name", "exon_number", "exon_id",
                           "level", "protein_id", "transcript_support_level",
                           "hgnc_id")

ref_annot_df <- separate(data = ref_annot_df, col = attribute, into = transcript_attributes,  sep = "; ")
for(header in transcript_attributes){
  ref_annot_df[,header] <- sub(".*? ", "",ref_annot_df[,header])
}


# create a dataframe that contains the exon that overlaps the breakpoint
df_breakpoints <- fusion_abdriged[,c("X.FusionName", "LeftGene", "RightGene", "LeftBreakpoint","RightBreakpoint")]
tmp <- gsub("\\..*", "", df_breakpoints$LeftGene) # removes version
df_breakpoints$LeftGene <- gsub(".*\\^", "", tmp) # remove gene symbol
tmp <- gsub("\\..*", "", df_breakpoints$RightGene) # removes version
df_breakpoints$RightGene <- gsub(".*\\^", "", tmp) # remove gene symbol
df_breakpoints <- separate(data = df_breakpoints, col = X.FusionName, into = c("LeftGene","RightGene"),  sep = "--")

df_fusion_annot <- as.data.frame(matrix(nrow=length(genes), ncol = 0))
df_fusion_annot$gene_name <- as.character(rbind(df_breakpoints$LeftGene, df_breakpoints$RightGene))
df_fusion_annot$genebreakpoints <- as.character(rbind(df_breakpoints$LeftBreakpoint, df_breakpoints$RightBreakpoint))
## get location
location <-sub('^([^:]+:[^:]+).*', '\\1', df_fusion_annot$genebreakpoints)
df_fusion_annot$chr <- gsub(":.*", "", location)
df_fusion_annot$pos <- gsub(".*:", "", location)
df_fusion_annot$pos <- as.integer(df_fusion_annot$pos)

tmp <- as.data.frame(matrix(nrow=0, ncol = ncol(ref_annot_df)))
colnames(tmp) <- colnames(ref_annot_df)
for(r in 1:nrow(df_fusion_annot)){
  dat  <- ref_annot_df[which(ref_annot_df$chr == df_fusion_annot[r,"chr"] & ref_annot_df$end >= df_fusion_annot[r,"pos"] & ref_annot_df$start <=  df_fusion_annot[r,"pos"]),]
  tmp <- rbind(dat,tmp)
}

tmp <- tmp[,c("source","chr", "start", "end", "strand", "gene_id", "transcript_id", 
              "gene_name", "hgnc_id", "transcript_name", "exon_number", "exon_id",
              "protein_id", "transcript_support_level")]


df_fusion_annot <- df_fusion_annot[,!(names(df_fusion_annot) %in% c("chr", "pos"))]
df2 <- df_fusion_annot %>% left_join( tmp, 
           by=c('gene_name'))


df2$MANE_Status <- mane$MANE_status[match(df2$transcript_id, mane$Ensembl_nuc)]

datatable(df2)
```

### 3.2 Fusion Prediction Annotation

```{r}
fusion_abdriged_annotation <- read.csv(params$fusionsabridgedcoding, sep = "\t")
fusion_abdriged_annotation <- fusion_abdriged_annotation[,c(1:5)]
genes <- unlist(strsplit(fusion_abdriged_annotation[,1], "--"))
```

```{r chimerkb}
chimerkb <- read_excel(params$chimerkb)
gene_nice <- gsub("--", "-", fusion_abdriged_annotation[,1])
chimerkb_sample_fusions <- chimerkb[which(chimerkb$Fusion_pair %in% gene_nice),]
datatable(chimerkb_sample_fusions)
```

```{r url_links}
fusion_abdriged_annotation$gene_fusion1 <- genes[seq_along(genes) %% 2 > 0]
fusion_abdriged_annotation$gene_fusion2 <- genes[seq_along(genes) %% 2 == 0]


make_hyperlink = function(myurl,mytext=myurl) {
  paste('<a href="',myurl,'">',mytext,'</a>')
}


fusion_abdriged_annotation$gene_fusion_cosmic <- ""
fusion_abdriged_annotation[1, "gene_fusion_cosmic"] <- "https://cancer.sanger.ac.uk/cosmic/fusion/overview?fid=322436&gid=368386"
fusion_abdriged_annotation[3, "gene_fusion_cosmic"] <- "https://cancer.sanger.ac.uk/cosmic/fusion/overview?fid=394618&gid=249070"


fusion_abdriged_annotation$gene_fusion1_oncokb <- c(paste("https://www.oncokb.org/gene/", fusion_abdriged_annotation$gene_fusion1, "/Fusions?refGenome=GRCh38" , sep = ""))
fusion_abdriged_annotation$gene_fusion2_oncokb <- c(paste("https://www.oncokb.org/gene/", fusion_abdriged_annotation$gene_fusion2, "/Fusions?refGenome=GRCh38" , sep = ""))

fusion_abdriged_annotation$gene_fusion1_CT <- c(paste("https://clinicaltrials.gov/search?term=", fusion_abdriged_annotation$gene_fusion1, "%20Fusion" , sep = ""))
fusion_abdriged_annotation$gene_fusion2_CT <- c(paste("https://clinicaltrials.gov/search?term=", fusion_abdriged_annotation$gene_fusion2, "%20Fusion" , sep = ""))


fusion_abdriged_annotation_final <-  fusion_abdriged_annotation %>%
  gt() %>%
  fmt(
    columns = c('gene_fusion_cosmic',
                'gene_fusion1_oncokb','gene_fusion2_oncokb',
                'gene_fusion1_CT','gene_fusion2_CT'),
    fns = make_hyperlink
  )

fusion_abdriged_annotation_final

```


### 3.3 Fusion Plots


```{r, warning=F,fig.width=10, fig.height=10}
FI_fusion = import_starfusion(params$fusionsabridgedcoding, "hg38")
plot_circle(FI_fusion)

```


```{r chimeraviz_plot_generation}

# get bam file
#fusion_reads <- "/home/aisha/Documents/Projects/cancer/sample_report/MB2002.star.bam"
#edb <- ensembldb::EnsDb("Homo_sapiens.GRCh38.103.sqlite" )


#for (fusion_number in 1:nrow(fusion_abdriged)){
#  fusion <- get_fusion_by_id(FI_fusion, fusion_number)
#  print(fusion_abdriged[fusion_number,1])
#  plot_fusion(
#    fusion = fusion,
#    bamfile = NULL,
#    edb = edb,
#    non_ucsc = T,
#    reduce_transcripts= F)
#  plot_fusion_transcript(
#    fusion,
#    edb)
#}
```

## 4. Session Info

```{r 4_session_info, include = TRUE}

sessionInfo()

```
